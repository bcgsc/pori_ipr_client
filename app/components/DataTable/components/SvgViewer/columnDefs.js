const columnDefs = [{
  colId: 'geneHeader',
  children: [
    { headerName: 'Position', field: 'position' },
    { headerName: 'HGNC Gene', field: 'gene' },
  ],
}, {
  colId: 'ensemblHeader',
  children: [
    { headerName: 'Ensembl Gene', field: 'ensemblGene', cellRenderer: 'LinkCellRenderer' },
    { headerName: 'Ensembl Transcript', field: 'ensemblTranscript', cellRenderer: 'LinkCellRenderer' },
  ],
}, {
  colId: 'predictedHeader',
  children: [
    { headerName: 'Exon', field: 'exon' },
    { headerName: 'Breakpoint', field: 'breakpoint' },
  ],
}];

export const setHeaderName = (header, colId) => {
  const index = columnDefs.findIndex((obj) => obj.colId === colId);
  columnDefs[index].headerName = header;
};

export default columnDefs;
