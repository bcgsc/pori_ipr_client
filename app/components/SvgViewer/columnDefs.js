
const columnDefs = [{
  colId: 'geneHeader',
  children: [
    { headerName: 'Position', field: 'position' },
    { headerName: 'HGNC Gene', field: 'gene' },
  ],
}, {
  colId: 'ensemblHeader',
  children: [
    { headerName: 'Ensembl Gene', field: 'ensemblGene' },
    { headerName: 'Ensembl Transcript', field: 'ensemblTranscript' },
  ],
}, {
  colId: 'predictedHeader',
  children: [
    { headerName: 'Exon', field: 'exon' },
    { headerName: 'Breakpoint', field: 'breakpoint' },
  ],
}];

export const setHeaderName = (header, colId) => {
  const index = columnDefs.findIndex(obj => obj.colId === colId);
  columnDefs[index].headerName = header;
};

export default columnDefs;
