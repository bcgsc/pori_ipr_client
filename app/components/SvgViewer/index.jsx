import React, {
  useEffect, useState, useRef, useCallback,
} from 'react';
import PropTypes from 'prop-types';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  Typography,
} from '@material-ui/core';
import { UncontrolledReactSVGPanZoom, fitToViewer } from 'react-svg-pan-zoom';
import InlineSVG from 'svg-inline-react';
import { AutoSizer } from 'react-virtualized';
import DataTable from '../DataTable';

/**
 * 
 * @param {*} props 
 */
function SvgViewer(props) {
  const {
    onClose,
    selectedRow,
    open,
  } = props;

  const [svgSize, setSvgSize] = useState({
    height: 500,
    width: 500,
  });

  const [viewer, setViewer] = useState();

  // useEffect(() => {
  //   if (viewer) {
  //     console.log(viewer);
  //     viewer.fitToViewer('left', 'top');
  //   }
  // }, [viewer]);

  useEffect(() => {
    if (selectedRow.svg) {
      const svg = new DOMParser().parseFromString(selectedRow.svg, 'image/svg+xml');
      const [svgElem] = svg.getElementsByTagName('svg');
      setSvgSize({
        height: svgElem.height.baseVal.value,
        width: svgElem.width.baseVal.value,
      });
    }
  }, [selectedRow]);

  const handleClose = (value) => {
    onClose(value);
  };

  const getFrameText = () => {
    switch (selectedRow.frame) {
      case 'OUT':
        return 'Out of frame';
      case 'IN':
        return 'In frame';
      case 'UNDET':
        return 'Not Determined';
      default:
        return 'N/A';
    }
  };

  return (
    <Dialog
      onClose={handleClose}
      open={open}
      maxWidth="xl"
      fullWidth
    >
      <DialogTitle>
        Structural Variant Details
      </DialogTitle>
      <DialogContent>
        {selectedRow.svg && (
          <AutoSizer disableHeight>
            {({ width }) => (
              <UncontrolledReactSVGPanZoom
                width={width}
                height={svgSize.height}
                background="#FFFFFF"
                detectAutoPan={false}
                customMiniature={() => null}
                toolbarProps={{ position: 'left' }}
                ref={setViewer}
              >
                <svg width={svgSize.width} height={svgSize.height}>
                  <InlineSVG src={selectedRow.svg} raw />
                </svg>
              </UncontrolledReactSVGPanZoom>
            )}
          </AutoSizer>
        )}
        <Typography>
          {selectedRow.svgTitle}
        </Typography>
      </DialogContent>
    </Dialog>
  );
}

SvgViewer.propTypes = {
  onClose: PropTypes.func.isRequired,
  selectedRow: PropTypes.objectOf(PropTypes.any).isRequired,
  open: PropTypes.bool.isRequired,
};

export default SvgViewer;
