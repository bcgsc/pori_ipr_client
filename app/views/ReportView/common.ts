import { ColDef } from '@ag-grid-community/core';

const sampleColumnDefs: ColDef[] = [
  {
    headerName: 'Sample',
    colId: 'sample',
    field: 'sample',
    hide: false,
  },
  {
    headerName: 'Sample Name',
    colId: 'sampleName',
    field: 'sampleName',
    hide: false,
  },
  {
    headerName: 'Collection Date',
    colId: 'collectionDate',
    field: 'collectionDate',
    hide: false,
  },
  {
    headerName: 'Primary Site',
    colId: 'primarySite',
    field: 'primarySite',
    hide: false,
  },
  {
    headerName: 'Biopsy Site',
    colId: 'biopsySite',
    field: 'biopsySite',
    hide: false,
  },
  {
    headerName: 'Pathology Estimated Tumour Content',
    colId: 'pathoTc',
    field: 'pathoTc',
    hide: false,
  },
];

export {
  sampleColumnDefs,
};
