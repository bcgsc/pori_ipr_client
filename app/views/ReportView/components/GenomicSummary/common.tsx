const VARIANT_TO_DISPLAY_NAME = {
  cnv: 'Copy Number Variants',
  smallMutation: 'Small Mutations',
  structuralVariant: 'Structural Variants',
  expression: 'Expression Outliers',
};

export {
  VARIANT_TO_DISPLAY_NAME,
};
