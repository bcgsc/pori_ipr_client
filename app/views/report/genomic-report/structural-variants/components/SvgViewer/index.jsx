import React from 'react';
import PropTypes from 'prop-types';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  Typography,
} from '@material-ui/core';

/**
 * 
 */
function SvgViewer(props) {
  const {
    open,
    selectedRow,
    onClose,
  } = props;

  const handleClose = (value) => {
    onClose(value);
  };

  return (
    <Dialog onClose={handleClose} open={open}>
      <DialogTitle>
        Fusion Diagram
      </DialogTitle>
    </Dialog>
  );
}

export default SvgViewer;
