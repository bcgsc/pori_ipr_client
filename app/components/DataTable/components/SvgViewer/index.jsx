import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  Button,
} from '@material-ui/core';
import { UncontrolledReactSVGPanZoom } from 'react-svg-pan-zoom';
import InlineSVG from 'svg-inline-react';
import AutoSizer from 'react-virtualized/dist/es/AutoSizer';
import DataTable from '../..';
import columnDefs, { setHeaderName } from './columnDefs';

import './index.scss';

const getFrameText = (frame) => {
  switch (frame) {
    case 'OUT':
      return 'Out of frame';
    case 'IN':
      return 'In frame';
    case 'UNDET':
      return 'Not Determined';
    default:
      return 'N/A';
  }
};

/**
 * @param {object} props props
 * @param {func} props.onClose parent close handler
 * @param {object} props.selectedRow current row object
 * @param {bool} props.open is open?
 * @return {*} JSX
 */
function SvgViewer(props) {
  const {
    onClose,
    selectedRow,
    open,
  } = props;

  const [svgSize, setSvgSize] = useState({
    height: 500,
    width: 500,
  });
  const [rowData, setRowData] = useState([]);

  useEffect(() => {
    if (selectedRow.svg) {
      const svg = new DOMParser().parseFromString(selectedRow.svg, 'image/svg+xml');
      const [svgElem] = svg.getElementsByTagName('svg');
      setSvgSize({
        height: svgElem.height.baseVal.value,
        width: svgElem.width.baseVal.value,
      });

      setRowData([{
        ...selectedRow,
        position: '5`',
        gene: selectedRow.gene1.name,
        ensemblGene: selectedRow.ntermGene,
        ensemblTranscript: selectedRow.ntermTranscript,
        exon: selectedRow.exon1,
        breakpoint: selectedRow.breakpoint.split('|')[0],
      }, {
        ...selectedRow,
        position: '3`',
        gene: selectedRow.gene2.name,
        ensemblGene: selectedRow.ctermGene,
        ensemblTranscript: selectedRow.ctermTranscript,
        exon: selectedRow.exon2,
        breakpoint: selectedRow.breakpoint.split('|')[1],
      }]);

      setHeaderName(`Gene: ${selectedRow.gene1.name} :: ${selectedRow.gene2.name}`, 'geneHeader');
      setHeaderName(`Type: ${selectedRow.eventType}`, 'ensemblHeader');
      setHeaderName(`Predicted: ${getFrameText(selectedRow.frame)}`, 'predictedHeader');
    }
  }, [selectedRow]);

  const handleClose = (value) => {
    onClose(value);
  };

  return (
    <Dialog
      onClose={handleClose}
      open={open}
      maxWidth="xl"
      fullWidth
    >
      <DialogTitle>
        <span className="dialog__title">
          <Typography variant="h5" align="center">
            Structural Variant Details
          </Typography>
        </span>
      </DialogTitle>
      <DialogContent>
        {selectedRow.svg && (
          <AutoSizer disableHeight>
            {({ width }) => (
              <UncontrolledReactSVGPanZoom
                width={width}
                height={svgSize.height}
                background="#FFFFFF"
                detectAutoPan={false}
                customMiniature={() => null}
                toolbarProps={{ position: 'left' }}
              >
                <svg width={svgSize.width} height={svgSize.height}>
                  <InlineSVG src={selectedRow.svg} raw />
                </svg>
              </UncontrolledReactSVGPanZoom>
            )}
          </AutoSizer>
        )}
        <DataTable
          columnDefs={columnDefs}
          rowData={rowData}
          canViewDetails={false}
        />
        <Typography>
          {selectedRow.svgTitle}
        </Typography>
      </DialogContent>
      <DialogActions>
        <Button color="primary" onClick={handleClose}>
          Close
        </Button>
      </DialogActions>
    </Dialog>
  );
}

SvgViewer.propTypes = {
  onClose: PropTypes.func.isRequired,
  selectedRow: PropTypes.objectOf(PropTypes.any).isRequired,
  open: PropTypes.bool.isRequired,
};

export default SvgViewer;
