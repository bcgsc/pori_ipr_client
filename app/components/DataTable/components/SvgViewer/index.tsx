import React, { useCallback, useEffect, useState } from 'react';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  Button,
} from '@mui/material';
import { AgGridReact } from '@ag-grid-community/react';

import SvgImage from '@/components/SvgImage';
import { StructuralVariantType } from '@/common';
import EnsemblCellRenderer from '../EnsemblCellRenderer';
import columnDefs, { setHeaderName } from './columnDefs';

import './index.scss';

const getFrameText = (frame: string): string => {
  switch (frame) {
    case 'OUT':
      return 'Out of frame';
    case 'IN':
      return 'In frame';
    case 'UNDET':
      return 'Not Determined';
    default:
      return 'N/A';
  }
};

type SvgViewerProps = {
  onClose: () => void;
  selectedRow: Partial<StructuralVariantType>;
  isOpen: boolean;
};

const SvgViewer = ({
  onClose,
  selectedRow,
  isOpen,
}: SvgViewerProps): JSX.Element => {
  const [rowData, setRowData] = useState([]);

  useEffect(() => {
    if (selectedRow.svg) {
      setRowData([{
        ...selectedRow,
        position: '5`',
        gene: selectedRow.gene1.name,
        ensemblGene: selectedRow.ntermGene,
        ensemblTranscript: selectedRow.ntermTranscript,
        exon: selectedRow.exon1,
        breakpoint: selectedRow.breakpoint.split('|')[0],
      }, {
        ...selectedRow,
        position: '3`',
        gene: selectedRow.gene2.name,
        ensemblGene: selectedRow.ctermGene,
        ensemblTranscript: selectedRow.ctermTranscript,
        exon: selectedRow.exon2,
        breakpoint: selectedRow.breakpoint.split('|')[1],
      }]);

      setHeaderName(`Gene: ${selectedRow.gene1.name} :: ${selectedRow.gene2.name}`, 'geneHeader');
      setHeaderName(`Type: ${selectedRow.eventType}`, 'ensemblHeader');
      setHeaderName(`Predicted: ${getFrameText(selectedRow.frame)}`, 'predictedHeader');
    }
  }, [selectedRow]);

  const handleClose = useCallback(() => {
    onClose();
  }, [onClose]);

  return (
    <Dialog
      onClose={handleClose}
      open={isOpen}
      maxWidth="lg"
      fullWidth
    >
      <DialogTitle>
        <span className="dialog__title">
          <Typography variant="h5" align="center">
            Structural Variant Details
          </Typography>
        </span>
      </DialogTitle>
      <DialogContent>
        {selectedRow.svg && (
          <SvgImage
            image={selectedRow.svg}
          />
        )}
        <div className="ag-theme-material">
          <AgGridReact
            columnDefs={columnDefs}
            rowData={rowData}
            domLayout="autoHeight"
            frameworkComponents={{
              EnsemblCellRenderer,
            }}
          />
        </div>
        <Typography variant="body2">
          {selectedRow.svgTitle}
        </Typography>
      </DialogContent>
      <DialogActions>
        <Button color="primary" onClick={handleClose}>
          Close
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default SvgViewer;
