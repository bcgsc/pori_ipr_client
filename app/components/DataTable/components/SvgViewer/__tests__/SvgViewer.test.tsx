import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { StructuralVariantType } from '@/common';
import SvgViewer from '..';

const mockSelectedRow = {
  breakpoint: 'na',
  gene1: {
    name: 'gene1',
  },
  gene2: {
    name: 'gene2',
  },
  svg: '<svg><g>test SVG</g></svg>',
  svgTitle: 'This is the text that appears',
} as Partial<StructuralVariantType>;

describe('SvgViewer', () => {
  test('It matches the snapshot', () => {
    const { asFragment } = render(
      <SvgViewer
        isOpen
        onClose={() => {}}
        selectedRow={mockSelectedRow}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('The dialog is shown when isOpen', async () => {
    render(
      <SvgViewer
        isOpen
        onClose={() => {}}
        selectedRow={mockSelectedRow}
      />,
    );
    expect(await screen.findAllByRole('presentation')).toHaveLength(2);
  });

  test('onClose is called', async () => {
    const mockOnClose = jest.fn();
    render(
      <SvgViewer
        isOpen
        onClose={mockOnClose}
        selectedRow={mockSelectedRow}
      />,
    );

    fireEvent.click(await screen.findByRole('button', { name: 'Close' }));
    expect(mockOnClose).toHaveBeenCalledTimes(1);
    expect(mockOnClose).toHaveBeenCalledWith();
  });

  test('The svgTitle text is shown', async () => {
    render(
      <SvgViewer
        isOpen
        onClose={() => {}}
        selectedRow={mockSelectedRow}
      />,
    );
    expect(await screen.findByText(mockSelectedRow.svgTitle)).toBeInTheDocument();
  });
});
