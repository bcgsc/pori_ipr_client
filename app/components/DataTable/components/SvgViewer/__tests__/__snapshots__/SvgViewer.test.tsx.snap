// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgViewer It matches the snapshot 1`] = `<DocumentFragment />`;
